### PR TITLE
PLANNER-2404 Avoid defensive copies of collections in ConstraintCollectors

### PR DESCRIPTION
**Pull request by triceo**

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java [hunk 1]
```diff
@@ -974,7 +977,10 @@ private static Runnable innerCountDistinctLong(CountDistinctLongResultContainer
      * Creates constraint collector that returns {@link Set} of the same element type as the {@link ConstraintStream}.
      * Makes no guarantees on iteration order.
      * For stable iteration order, use {@link #toSortedSet()}.
-     *
+     * <p>
+     * The contents of the collection will be changing during {@link org.optaplanner.core.api.solver.Solver} lifetime,
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java:981]:
```suggestion
     * The contents of the collection will change during {@link org.optaplanner.core.api.solver.Solver} lifetime,
```

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java [hunk 2]
```diff
@@ -985,18 +991,24 @@ private static Runnable innerCountDistinctLong(CountDistinctLongResultContainer
     /**
      * Creates constraint collector that returns {@link SortedSet} of the same element type as the
      * {@link ConstraintStream}.
+     * <p>
+     * The contents of the collection will be changing during {@link org.optaplanner.core.api.solver.Solver} lifetime,
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java:995]:
```suggestion
     * The contents of the collection will change during {@link org.optaplanner.core.api.solver.Solver} lifetime,
```

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java [hunk 3]
```diff
@@ -985,18 +991,24 @@ private static Runnable innerCountDistinctLong(CountDistinctLongResultContainer
...
-        return toCollection(Function.identity(), i -> new TreeSet<>());
+        return toSortedSet(a -> a);
     }
 
     /**
      * Creates constraint collector that returns {@link List} of the same element type as the {@link ConstraintStream}.
      * Makes no guarantees on iteration order.
      * For stable iteration order, use {@link #toSortedSet()}.
+     * <p>
+     * The contents of the collection will be changing during {@link org.optaplanner.core.api.solver.Solver} lifetime,
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java:1010]:
```suggestion
     * The contents of the collection will change during {@link org.optaplanner.core.api.solver.Solver} lifetime,
```

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java [hunk 4]
```diff
@@ -985,18 +991,24 @@ private static Runnable innerCountDistinctLong(CountDistinctLongResultContainer
     /**
      * Creates constraint collector that returns {@link SortedSet} of the same element type as the
      * {@link ConstraintStream}.
+     * <p>
+     * The contents of the collection will be changing during {@link org.optaplanner.core.api.solver.Solver} lifetime,
+     * they must therefore never be modified nor stored without making a defensive copy first.
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java:996]: Can the users modify or store elements of that list? No, they can't.
In a vanilla CS, can the users modify or store the tuple facts? No, they can't.

I don't think this needs to be explicitly stated.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java [hunk 5]
```diff
@@ -1406,6 +1524,8 @@ public void remove(Key key, Value value) {
      * <p>
      * Makes no guarantees on iteration order for map entries.
      * For stable iteration order, use {@link #toSortedMap(Function, Function, BinaryOperator)}.
+     * <p>
+     * This constraint collector is slow and causes considerable garbage collector pressure.
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java:1528]: If we do this change for toList and toSet and toSortedSet, we should do it for toMap etc too, for consistency (principle of least surprise).

And yes, I think we need to do this change :)

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java [hunk 6]
```diff
@@ -1225,13 +1331,18 @@ private static Runnable innerCountDistinctLong(CountDistinctLongResultContainer
      */
     public static <A, B, C, D, Mapped extends Comparable<Mapped>> QuadConstraintCollector<A, B, C, D, ?, SortedSet<Mapped>>
             toSortedSet(QuadFunction<A, B, C, D, Mapped> groupValueMapping) {
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java:1333]: Separate issue: as we support toSortedSet(mapping), we should probably also support toSortedSet(mapping, Comparator). Not having it is an API gap.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java (hunk 5 above)
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java:1528]: Making `toMap()` incremental in this way is likely impossible.
The interface of the method requires post-processing on the internal map before the collector returns.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java (hunk 6 above)
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java:1333]: Making toMap() incremental in this way is likely impossible.
The interface of the method requires post-processing on the internal map before the collector returns.

**Review by ge0ffrey — APPROVED** [review]
Great improvement!

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java (hunk 6 above)
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java:1333]: https://issues.redhat.com/browse/PLANNER-2405

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java (hunk 5 above)
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java:1528]: https://issues.redhat.com/browse/PLANNER-2406

**Pull request merged**